core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala

```diff
@@ -0,0 +1,170 @@
+import collection.JavaConversions._
+
+case class CoordinatorEpochAndMarkers(metadataPartition: Int, coordinatorEpoch: Int, txnMarkerEntry: util.List[WriteTxnMarkersRequest.TxnMarkerEntry])
+case class DestinationBrokerAndQueuedMarkers(destBrokerNode: Node, markersQueue: BlockingQueue[CoordinatorEpochAndMarkers])
+
+object TransactionMarkerChannelManager {
+  def apply(config: KafkaConfig,
+            metrics: Metrics,
+            metadataCache: MetadataCache,
+            txnMarkerPurgatory: DelayedOperationPurgatory[DelayedTxnMarker],
+            time: Time): TransactionMarkerChannelManager = {
+
+    val channel = new TransactionMarkerChannel(config.interBrokerListenerName, metadataCache)
+    val sendThread: InterBrokerSendThread = {
+      val threadName = "TxnMarkerSenderThread-" + config.brokerId
+      val channelBuilder = ChannelBuilders.clientChannelBuilder(
+        config.interBrokerSecurityProtocol,
+        JaasContext.Type.SERVER,
+        config,
+        config.interBrokerListenerName,
+        config.saslMechanismInterBrokerProtocol,
+        config.saslInterBrokerHandshakeRequestEnable
+      )
+      val selector = new Selector(
+        NetworkReceive.UNLIMITED,
+        config.connectionsMaxIdleMs,
+        metrics,
+        time,
+        "txn-marker-channel",
+        Map("broker-id" -> config.brokerId.toString).asJava,
+        false,
+        channelBuilder
+      )
+      val networkClient = new NetworkClient(
+        selector,
+        new ManualMetadataUpdater(),
+        threadName,
+        1,
+        50,
+        Selectable.USE_DEFAULT_BUFFER_SIZE,
+        config.socketReceiveBufferBytes,
+        config.requestTimeoutMs,
+        time,
+        false,
+        new ApiVersions
+      )
+
+      new InterBrokerSendThread(threadName, networkClient, requestGenerator(channel, txnMarkerPurgatory), time)
+    }
+
+    new TransactionMarkerChannelManager(config,
+      metadataCache,
+      txnMarkerPurgatory,
+      sendThread,
+      channel)
+  }
+
+
+  private[transaction] def requestGenerator(transactionMarkerChannel: TransactionMarkerChannel,
+                                            txnMarkerPurgatory: DelayedOperationPurgatory[DelayedTxnMarker]): () => Iterable[RequestAndCompletionHandler] = {
+    def generateRequests(): Iterable[RequestAndCompletionHandler] = {
+      transactionMarkerChannel.brokerStateMap.flatMap {case (brokerId: Int, destAndMarkerQueue: DestinationBrokerAndQueuedMarkers) =>
+        val markersToSend: java.util.List[CoordinatorEpochAndMarkers] = new util.ArrayList[CoordinatorEpochAndMarkers] ()
+        destAndMarkerQueue.markersQueue.drainTo (markersToSend)
+        markersToSend.groupBy{ epochAndMarker => (epochAndMarker.metadataPartition, epochAndMarker.coordinatorEpoch) }
+          .map { case((metadataPartition: Int, coordinatorEpoch:Int), buffer: mutable.Buffer[CoordinatorEpochAndMarkers]) =>
+            val txnMarkerEntries = buffer.flatMap { x => x.txnMarkerEntry }.asJava
```

Comment:
`flatMap` is preferable to `map` and then `flatten`. I am guessing Jun's question is because `txnMarkerEntry` sounds like an element, but it's actually a `List`. Maybe it should be renamed?